### DISCUSSION TITLE
Version 2.1.4

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#221] [Announcements]
## General

1. MuJoCo now uses [GLAD](https://github.com/Dav1dde/glad) to manage OpenGL API access instead of GLEW. On Linux, there is no longer a need to link against different GL wrangling libraries depending on whether GLX, EGL, or OSMesa is being used. Instead, users can simply use GLX, EGL, or OSMesa to create a GL context and `mjr_makeContext` will detect which one is being used.

2. Add visualisation for contact frames. This is useful when writing or modifying collision functions, when the actual direction of the x and y axes of a contact can be important.

## Binary build

3. The `_nogl` dynamic library is no longer provided on Linux and Windows. The switch to GLAD allows us to resolve OpenGL symbols when `mjr_makeContext` is called rather than when the library is loaded. As a result, the MuJoCo library no longer has an explicit dynamic dependency on OpenGL, and can be used on system where OpenGL is not present.

## Simulate

4. Fix a bug in simulate where pressing '[' or ']' when a model is not loaded causes a crash.

5. Contact frame visualisation is added to the Simulate GUI.

6. Rename "set key", "reset to key" to "save key" and "load key", respectively.

7. Change bindings of F6 and F7 from the not very useful "vertical sync" and "busy wait" to the more useful cycling of frames and labels.

## Bug fixes

8. `mj_resetData` zeroes out the `solver_nnz` field.

9. Remove a special branch in `mju_quat2mat` for unit quaternions. Previously, `mju_quat2mat` skipped all computation if the real part of the quaternion equals 1.0. For very small angles (e.g. when finite differencing), the cosine can evaluate to exactly 1.0 at double precision while the sine is still nonzero.

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.1.4'>2.1.4</a>.</em>

**Comment by mhming** [comment]
There was a small mistake in windows version **mujoco-2.1.4-windows-x86_64.zip** , GLFW in sample->Makefile  should be **glfw3dll.lib**, this leed a Link error!

COMMON=/O2 /MT /EHsc /arch:AVX /I../include /Fe../bin/

all:
	cl $(COMMON) testxml.cc                 ../lib/mujoco.lib
	cl $(COMMON) testspeed.cc               ../lib/mujoco.lib
	cl $(COMMON) compile.cc                 ../lib/mujoco.lib
	cl $(COMMON) derivative.cc /openmp      ../lib/mujoco.lib
	cl $(COMMON) basic.cc ../lib/glfw3dll.lib  ../lib/mujoco.lib
	cl $(COMMON) record.cc ../lib/glfw3dll.lib ../lib/mujoco.lib
	cl $(COMMON) simulate.cc uitools.c ../lib/glfw3dll.lib ../lib/mujoco.lib
	del *.obj

**Reply by saran-t** [reply]
The comment on top of the Makefile suggests that `glfw3dll.lib` should be copied and renamed to `glfw3.lib`, but yes you're right, we should just modify the Makefile itself to use `glfw3dll.lib` instead!

**Reply by mhming** [reply]
Thanks for your job, change to glfw3dll.lib instead will be much clearer

**Reply by saran-t** [reply]
The Makefile that comes bundled with the 2.1.5 release has been fixed.
